### ISSUE TITLE
Failed Launch when deploying Rails App (Digital Ocean and AWS)

### ISSUE THREAD

**Issue by choncou** [flynn/flynn#3426]
I have tried deploying a Rails app with Flynn and it failed on both AWS and Digital Ocean. 
Here is the error that occurred in both cases.
```
Running: rake assets:precompile
       rake aborted!
       ArgumentError: Missing required arguments: aws_access_key_id, aws_secret_access_key
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/fog-core-1.42.0/lib/fog/core/service.rb:244:in `validate_options'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/fog-core-1.42.0/lib/fog/core/service.rb:268:in `handle_settings'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/fog-core-1.42.0/lib/fog/core/service.rb:98:in `new'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/fog-core-1.42.0/lib/fog/core/services_mixin.rb:16:in `new'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/fog-core-1.42.0/lib/fog/storage.rb:27:in `new'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/carrierwave-0.11.2/lib/carrierwave/uploader/configuration.rb:83:in `eager_load_fog'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/carrierwave-0.11.2/lib/carrierwave/uploader/configuration.rb:96:in `fog_credentials='
       /tmp/build/config/initializers/carrier_wave.rb:3:in `block in <top (required)>'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/carrierwave-0.11.2/lib/carrierwave/uploader/configuration.rb:118:in `configure'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/carrierwave-0.11.2/lib/carrierwave.rb:14:in `configure'
       /tmp/build/config/initializers/carrier_wave.rb:2:in `<top (required)>'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/dependencies.rb:287:in `load'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/dependencies.rb:287:in `block in load'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/dependencies.rb:259:in `load_dependency'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/dependencies.rb:287:in `load'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/engine.rb:648:in `block in load_config_initializer'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/activesupport-5.0.0.1/lib/active_support/notifications.rb:166:in `instrument'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/engine.rb:647:in `load_config_initializer'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/engine.rb:612:in `block (2 levels) in <class:Engine>'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/engine.rb:611:in `each'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/engine.rb:611:in `block in <class:Engine>'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/initializable.rb:30:in `instance_exec'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/initializable.rb:30:in `run'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/initializable.rb:55:in `block in run_initializers'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/initializable.rb:44:in `each'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/initializable.rb:44:in `tsort_each_child'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/initializable.rb:54:in `run_initializers'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/application.rb:352:in `initialize!'
       /tmp/build/config/environment.rb:5:in `<top (required)>'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/application.rb:328:in `require'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/application.rb:328:in `require_environment!'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/railties-5.0.0.1/lib/rails/application.rb:448:in `block in run_tasks_blocks'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/sprockets-rails-3.1.1/lib/sprockets/rails/task.rb:62:in `block (2 levels) in define'
       /tmp/build/vendor/bundle/ruby/2.3.0/gems/rake-11.2.2/exe/rake:27:in `<top (required)>'
       Tasks: TOP => environment
       (See full trace by running task with --trace)
       !
       !     Precompiling assets failed.
       !
ERROR: Build failed: exec: job exited with status 1
exit status 1
```

I set the credentials which it shows are missing inside a file: _project_root/config/initializers/carrier_wave.rb_

```
if Rails.env.production?
  CarrierWave.configure do |config|
    config.fog_credentials = {
      # Configuration for Amazon S3
      provider: 'AWS',
      aws_access_key_id: ENV['S3_ACCESS_KEY'],
      aws_secret_access_key: ENV['S3_SECRET_KEY'],
      region: ENV['S3_REGION']
    }
    config.fog_directory = ENV['S3_BUCKET']
  end
end
```

And just before clicking launch on the Flynn dashboard, I enter all the required Environment Variables. I have this project working perfectly on heroku. Any idea if I am doing anything wrong?

**Comment by josephglanville** [comment]
You may need to check the value of "RAILS_ENV" to ensure that `Rails.env.production?` is returning true and setting the CarrierWave settings.

**Comment by choncou** [comment]
I reattempted the launch and set `RAILS_ENV` to production. It still ran into the same error.

**Comment by lmars** [comment]
@choncou can you see the `S3_*` environment variables when you run `flynn env` from the command line?

**Comment by choncou** [comment]
@lmars yes the environment variable do appear. And I also see that my app appears on my flynn cloud dashboard. When I visit the route that has been set [here](https://choncou-twitterclone-master.o6l5.flynnhub.com/). It shows "Service Unavailable".

**Comment by lmars** [comment]
@choncou I can't seem to reproduce this. I get the same error as you when the environment variables are not set, but as soon as I set them the deploy goes through ok:

```
$ cat config/initializers/carrierwave.rb 
if Rails.env.production?
  CarrierWave.configure do |config|
    config.fog_credentials = {
      # Configuration for Amazon S3
      provider: 'AWS',
      aws_access_key_id: ENV['S3_ACCESS_KEY'],
      aws_secret_access_key: ENV['S3_SECRET_KEY'],
      region: ENV['S3_REGION']
    }
    config.fog_directory = ENV['S3_BUCKET']
  end
end

$ flynn env

$ git push flynn master
...
-----> Compiling Ruby/Rails
...
-----> Preparing app for Rails asset pipeline
       Running: rake assets:precompile
       rake aborted!
       ArgumentError: Missing required arguments: aws_access_key_id, aws_secret_access_key
...

$ flynn env set S3_ACCESS_KEY=xxx S3_SECRET_KEY=xxx
Created release 06ed4dad-4c98-4ef7-b79e-5f53cd526751.

$ git push flynn master
...
-----> Preparing app for Rails asset pipeline
       Running: rake assets:precompile
...
       Asset precompilation completed (3.66s)
       Cleaning assets
       Running: rake assets:clean
...
=====> Application deployed
```

Can you try deploying from the command line in this way and see if it works? If it does, then this is probably a bug in the dashboard, so it would be useful if you could give us the exact steps you are taking to deploy so we can try and reproduce.

**Comment by choncou** [comment]
@lmars It seems to have worked perfectly through command line and no longer showed the same errors. The app is now live. Thanks!

The steps I followed to launch through the dashboard:

1. Connected my Github account
2. Clicked `Create Application`
3. Selected my rails project from the list of GitHub repos
4. Clicked launch on my latest Master branch commit
5. Selected PostgreSQL and Redis
6. Entered in the S3 environment variables

After that I watched the logs during deployment until it ran into the original error I posted.
